File: packages/ik_rules/lib/src/equipment/specialist.dart
import 'package:ik_content/ik_content.dart';

import '../config.dart';
import '../save/generated/save_models.dart';

const String essenceItemId = 'ITEM-0011';
const String chefHatItemId = 'ITEM-0165';
const String wizardHatItemId = 'ITEM-0166';
const String quiverItemId = 'ITEM-0303';
const String alchemistGogglesItemId = 'ITEM-0318';
const String cookingSkillId = 'SKL-0007';
const String huntingSkillId = 'SKL-0005';
const String alchemySkillId = 'SKL-0010';

const double chefHatDoubleChance = 1 / 100;
const double wizardHatEssenceFactor = 0.99;
const double quiverHuntingXpFactor = 1.05;

/// Max potions granted per Alchemy craft (queue space assumes this).
const num alchemyPotionOutputMax = 3;

/// Min potions granted per Alchemy craft without Alchemist Goggles.
const num alchemyPotionOutputMin = 1;

/// Min potions granted per Alchemy craft with Alchemist Goggles.
const num alchemyPotionOutputGogglesMin = 2;

String specialistChefHatItemId(GameDatabase db) =>
    configString(db, 'specialist.chef_hat_item_id', chefHatItemId);

String specialistWizardHatItemId(GameDatabase db) =>
    configString(db, 'specialist.wizard_hat_item_id', wizardHatItemId);

String specialistQuiverItemId(GameDatabase db) =>
    configString(db, 'specialist.quiver_item_id', quiverItemId);

String specialistAlchemistGogglesItemId(GameDatabase db) =>
    configString(db, 'specialist.alchemist_goggles_item_id', alchemistGogglesItemId);

bool hasEquippedItem(PlayerSave save, String itemId) {
  return save.equipment.slots.values.any((stack) => stack?.itemId == itemId);
}

/// 1% essence discount, with the remaining cost rounded up.
num wizardEssenceCost(GameDatabase db, num baseQuantity, PlayerSave save) {
  if (baseQuantity <= 0) return 0;
  if (!hasEquippedItem(save, specialistWizardHatItemId(db))) return baseQuantity;
  final factor = configNumber(db, 'specialist.wizard_hat_essence_factor', wizardHatEssenceFactor);
  return (baseQuantity * factor).ceil();
}

num applyQuiverHuntingXp(GameDatabase db, num amount, PlayerSave save, String skillId) {
  if (amount <= 0) return 0;
  if (skillId != huntingSkillId) return amount;
  if (!hasEquippedItem(save, specialistQuiverItemId(db))) return amount;
  final factor = configNumber(db, 'specialist.quiver_hunting_xp_factor', quiverHuntingXpFactor);
  return (amount * factor).floor();
}

num chefHatOutputQuantity(
  GameDatabase db,
  num baseQuantity,
  PlayerSave save,
  String skillId,
  double Function() random,
) {
  if (baseQuantity <= 0) return 0;
  if (skillId != cookingSkillId) return baseQuantity;
  if (!hasEquippedItem(save, specialistChefHatItemId(db))) return baseQuantity;
  final chance = configNumber(db, 'specialist.chef_hat_double_chance', chefHatDoubleChance);
  if (random() >= chance) return baseQuantity;
  return baseQuantity * 2;
}

/// Fair die for potion crafts: 1–3 normally, 2–3 with Alchemist Goggles.
num alchemyPotionOutputQuantity(
  GameDatabase db,
  num baseQuantity,
  PlayerSave save,
  String skillId,
  double Function() random,
) {
  if (baseQuantity <= 0) return 0;
  if (skillId != alchemySkillId) return baseQuantity;
  final roll = random();
  final max = configNumber(db, 'specialist.alchemy_potion_output_max', alchemyPotionOutputMax);
  final min = configNumber(db, 'specialist.alchemy_potion_output_min', alchemyPotionOutputMin);
  final gogglesMin = configNumber(
    db,
    'specialist.alchemy_potion_output_goggles_min',
    alchemyPotionOutputGogglesMin,
  );
  if (hasEquippedItem(save, specialistAlchemistGogglesItemId(db))) {
    final die = (roll * (max - gogglesMin + 1)).floor() + gogglesMin;
    return baseQuantity * die;
  }
  final die = (roll * max).floor() + min;
  return baseQuantity * die;
}

/// Per-craft output used when reserving bag space for a production queue.
num productionOutputReservePerCraft(GameDatabase db, String skillId, num baseQuantity) {
  if (skillId == alchemySkillId) {
    return baseQuantity *
        configNumber(db, 'specialist.alchemy_potion_output_max', alchemyPotionOutputMax);
  }
  return baseQuantity;
}

File: src/game/equipment/specialist.ts
import type { PlayerSave } from '../save/types'

export const ESSENCE_ITEM_ID = 'ITEM-0011'
export const CHEF_HAT_ITEM_ID = 'ITEM-0165'
export const WIZARD_HAT_ITEM_ID = 'ITEM-0166'
export const QUIVER_ITEM_ID = 'ITEM-0303'
export const ALCHEMIST_GOGGLES_ITEM_ID = 'ITEM-0318'
export const COOKING_SKILL_ID = 'SKL-0007'
export const HUNTING_SKILL_ID = 'SKL-0005'
export const ALCHEMY_SKILL_ID = 'SKL-0010'

export const CHEF_HAT_DOUBLE_CHANCE = 1 / 100
export const WIZARD_HAT_ESSENCE_FACTOR = 0.99
export const QUIVER_HUNTING_XP_FACTOR = 1.05
/** Max potions granted per Alchemy craft (queue space assumes this). */
export const ALCHEMY_POTION_OUTPUT_MAX = 3
/** Min potions granted per Alchemy craft without Alchemist Goggles. */
export const ALCHEMY_POTION_OUTPUT_MIN = 1
/** Min potions granted per Alchemy craft with Alchemist Goggles. */
export const ALCHEMY_POTION_OUTPUT_GOGGLES_MIN = 2

export function hasEquippedItem(save: PlayerSave, itemId: string): boolean {
  return Object.values(save.equipment.slots).some((stack) => stack?.itemId === itemId)
}

/** 1% essence discount, with the remaining cost rounded up. */
export function wizardEssenceCost(baseQuantity: number, save: PlayerSave): number {
  if (baseQuantity <= 0) return 0
  if (!hasEquippedItem(save, WIZARD_HAT_ITEM_ID)) return baseQuantity
  return Math.ceil(baseQuantity * WIZARD_HAT_ESSENCE_FACTOR)
}

export function applyQuiverHuntingXp(amount: number, save: PlayerSave, skillId: string): number {
  if (amount <= 0) return 0
  if (skillId !== HUNTING_SKILL_ID) return amount
  if (!hasEquippedItem(save, QUIVER_ITEM_ID)) return amount
  return Math.floor(amount * QUIVER_HUNTING_XP_FACTOR)
}

export function chefHatOutputQuantity(
  baseQuantity: number,
  save: PlayerSave,
  skillId: string,
  random: () => number,
): number {
  if (baseQuantity <= 0) return 0
  if (skillId !== COOKING_SKILL_ID) return baseQuantity
  if (!hasEquippedItem(save, CHEF_HAT_ITEM_ID)) return baseQuantity
  if (random() >= CHEF_HAT_DOUBLE_CHANCE) return baseQuantity
  return baseQuantity * 2
}

/** Fair die for potion crafts: 1–3 normally, 2–3 with Alchemist Goggles. */
export function alchemyPotionOutputQuantity(
  baseQuantity: number,
  save: PlayerSave,
  skillId: string,
  random: () => number,
): number {
  if (baseQuantity <= 0) return 0
  if (skillId !== ALCHEMY_SKILL_ID) return baseQuantity
  const roll = random()
  if (hasEquippedItem(save, ALCHEMIST_GOGGLES_ITEM_ID)) {
    const die = Math.floor(roll * 2) + ALCHEMY_POTION_OUTPUT_GOGGLES_MIN
    return baseQuantity * die
  }
  const die = Math.floor(roll * ALCHEMY_POTION_OUTPUT_MAX) + ALCHEMY_POTION_OUTPUT_MIN
  return baseQuantity * die
}

/** Per-craft output used when reserving bag space for a production queue. */
export function productionOutputReservePerCraft(
  skillId: string,
  baseQuantity: number,
): number {
  if (skillId === ALCHEMY_SKILL_ID) {
    return baseQuantity * ALCHEMY_POTION_OUTPUT_MAX
  }
  return baseQuantity
}
